doc/developer/design/20240311_adapter_catalog_listen.md

```diff
@@ -0,0 +1,178 @@
+}
+```
+
+### Durable Catalog API
+
+```Rust
+trait ReadOnlyDurableCatalogState {
+    // ...
+
+    /// Fetches and returns all updates to the durable catalog state that have not yet been
+    /// consumed.
+    async fn sync(&mut self) -> Vec<StateUpdate>;
+
+    // ...
+}
+```
+
+### In-memory Catalog API
+
+```Rust
+impl Catalog {
+    /// Fetches and applies all updates to the durable catalog state that have not yet been
+    /// consumed.
+    async fn sync(&mut self) -> (Vec<BuiltinTableUpdate>, Vec<ControllerCommand>) {
+        let updates = self.storage().await.sync().await;
+        self.apply_updates(updates);
+    }
+
+    /// Update in-memory catalog state and generate builtin table updates and controller commands. 
+    fn apply_updates(&mut self, updates: Vec<StateUpdate>) -> (Vec<BuiltinTableUpdate>, Vec<ControllerCommand>) {
+        let mut builtin_table_updates = Vec::new();
+        let mut controller_commands = Vec::new();
+        for StateUpdate { kind, diff } in updates {
+            match (kind, diff) {
+                // ...
+            }
+        }
+        (builtin_table_updates, controller_commands)
+    }
+}
+```
+
+### Coordinator API
+
+```Rust
+impl Coordinator {
+    /// Appends updates to builtin tables.
+    async fn apply_builtin_table_updates(builtin_table_updates: Vec<BuiltinTableUpdate>) {
+        // ...
+    }
+
+    /// Sends commands to the controller.
+    async fn apply_controller_commands(controller_commands: Vec<ControllerCommand>) {
+        // ...
+    }
+}
+```
+
+### Coordinator Bootstrap Psuedo Code
+
+Note: This incorrectly simplifies the current structure of the code and combines some logic from
+`mz_adapter::coord::serve`, `Catalog::open`, and `Coordinator::bootstrap` for ease of reading.
+
+```Rust
+impl Coordinator {
+    async fn serve(&mut self) {
+        // ... 
+        let durable_catalog_state = self.open_durable_catalog_state().await;
+        let catalog = Catalog::new(durable_catalog_state);
+        // Get retractions of current builtin table contents.  
+        let mut builtin_table_updates = self.get_retractions_for_all_builtin_table_contents();
+        let (new_builtin_table_updates, controller_commands) = catalog.sync().await;
+        builtin_table_updates.extend(new_builtin_table_updates);
+        self.apply_builtin_table_updates(builtin_table_updates).await;
+        self.apply_controller_commands(controller_commands).await;
+        // ...
+    }
+}
+```
+
+### Coordinator Sequence Psuedo Code
+
+```Rust
+impl Coordinator {
+    pub(crate) async fn sequence_plan(
+        &mut self,
+        mut ctx: ExecuteContext,
+        plan: Plan,
+        resolved_ids: ResolvedIds,
+    ) -> LocalBoxFuture<'_, ()> {
+        // ...
+        // IMPORTANT: This may trigger a catalog transaction which only updates the durable catalog.
+        self.sequence_X(plan);
+        // ...
+        let (builtin_table_updates, controller_commands) = self.catalog.sync().await;
```

Comment:
So this `sync` could potentially get two things worth of updates depending on races, but again, that's fine because we're not doing "Multi-subscriber" yet

Reply:
Can you expand on this a bit? The race here is that multiple subscribers will get the exact same set of `builtin_table_updates` and `controller_commands`. If they all try and apply them, then we'll end up with duplicate rows in the builtin tables and who knows what in storage/compute. However, there shouldn't be a race in updating the state of `catalog`. I.e. all subscribers should have the exactly correct state of objects in `catalog`.

Reply:
Ah, it's possible that the race I was imagining doesn't exist, but I was picturing two pgwire sessions concurrently running e.g. a `CREATE TABLE`. They'd write to the durable catalog in some order, but who knows in what order they'd call `sync`

Reply:
It is possible that two catalogs are arbitrarily ahead/behind each other. However, they will observe the changes to the persist shard in the same order. For example consider the following orderings:

----

1. `A` creates table `t1`.
2. `B` creates table `t2`.
3. `A` calls `sync` and observes `t1` and `t2`.
4. `B` calls `sync` and observes `t1` and `t2`.

----

1. `A` creates table `t1`.
2. `B` creates table `t2`.
3. `B` calls `sync` and observes `t1` and `t2`.
4. `A` calls `sync` and observes `t1` and `t2`.

----

1. `A` creates table `t1`.
2. `A` calls `sync` and observes `t1`.
3. `B` creates table `t2`.
5. `B` calls `sync` and observes `t1` and `t2`.

----

etc.

Reply:
I think maybe the piece that is not clear from my doc is that calling `catalog.sync()` has no effect on other catalogs. Which events have already been consumed or not will be stored entirely in memory and not durably written down anywhere.

Reply:
> So this sync could potentially get two things worth of updates depending on races, but again, that's fine because we're not doing "Multi-subscriber" yet

> 1. `A` creates table `t1`.
> 2. `B` creates table `t2`.
> 3. `A` calls `sync` and observes `t1` and `t2`.
> 4. `B` calls `sync` and observes `t1` and `t2`.

> 1. `A` creates table `t1`.
> 2. `A` calls `sync` and observes `t1`.
> 3. `B` creates table `t2`.
> 4. `B` calls `sync` and observes `t1` and `t2`.

Heh, I think we're just talking past each other. The ABAB example (and the ABBA one I didn't quote here) is what I meant by "sync could potentially get two things worth of updates". And this is fine because you've defined more than one thing handling apply_builtin_table_updates/apply_controller_commands as out of scope

Sorry, I wasn't pointing out a flaw, just talking myself through how it worked :)

Reply:
Just to check my own understanding, as you've written it, there's no possibility of concurrency between a call to `sequence` and a call to `sync`, right? That's all happening on the coord main loop, right? So you're guaranteed that every `sequence_foo` will be immediately followed by a call to `sync`, without the possibility of the coordinator running off and sequencing another command?

Reply:
> Just to check my own understanding, as you've written it, there's no possibility of concurrency between a call to sequence and a call to sync, right? That's all happening on the coord main loop, right? So you're guaranteed that every sequence_foo will be immediately followed by a call to sync, without the possibility of the coordinator running off and sequencing another command?

Ah good question/point. No, there is no guarantee anymore that `sequence_X` will run to completion without the possibility of the coordinator running off and sequencing another command. I think in practice none of the DDL perform off-thread work after calling `catalog_transact`. Still, that's a bit scary to rely on. I think we should move this logic into `catalog_transact`. Something like


```Rust
impl Coordinator {
  fn catalog_transact(ops: Vec<Op>, ...) {
    // Execute catalog transaction updating the durable catalog only.
    let (builtin_table_updates, controller_commands) = self.catalog.sync().await;
    self.apply_builtin_table_updates(builtin_table_updates).await;
    self.apply_controller_commands(controller_commands).await;
  }
}
```

Reply:
> Ah, because they're all `async`?

Not exactly. Many of the `sequence` functions are structured in a way that they create new tasks and perform some work in that task while allowing the Coordinator to go and process new messages. For example, `CREATE MATERIALIZED VIEW` does this with it's optimization stage, https://github.com/MaterializeInc/materialize/blob/38eabd8146f3609ff671cd9319d31108398ad0d5/src/adapter/src/coord/sequencer/inner/create_materialized_view.rs#L420-L499 (I was wrong about DDL not doing off task work).

> What if we made the sequence functions return `Vec<Op>`?

For the same reason listed above, we can't have the `sequence` functions return anything meaningful. They return early before they're complete and finish their work in a new task that eventually sends a message back to the Coordinator.

I'll have to have a think about the best way of doing this. An even more intermediate step would be to ignore the builtin table updates and controller commands and just focus on having `Catalog` update it's in memory state in `Catalog::sync`. Then I think that logic can live in `catalog_transact` and we can kick the can on everything else.

Reply:
If builtin table updates and controller commands are not necessary for a multi-subscriber catalog, then maybe it's more prudent to ignore them for now.

Reply:
> If builtin table updates and controller commands are not necessary for a multi-subscriber catalog, then maybe it's more prudent to ignore them for now.

I think it'd be good to handle them! Even though we don't know how to _apply_ builtin table updates from multiple live environments today, that's exactly the thing we'll want to solve for full UCI and 0DT. So having the catalog API expose them will be really helpful, I think.

Your idea of just doing it all in `catalog_transact` sounds good to me. There's still the potential for misuse, but cest le vie. We can just enforce the code structure rules (only one call to `catalog_transact`; don't do anything after) by convention and comments for now.